File: e-learning-platform/nodejs/routes/quiz.js
const express = require('express');
const router = express.Router();
const { Question, QuizResult } = require('../models');
const { calculateScore, calculatePercentage, isPassed } = require('../services/scoreCalculator');
const { auth, requireRole } = require('../middleware/auth');

router.get('/quiz', auth, requireRole('STUDENT'), async (req, res) => {
  const questions = await Question.findAll();
  res.render('quiz', { questions });
});

router.post('/quiz/submit', auth, requireRole('STUDENT'), async (req, res) => {
  const questions = await Question.findAll();
  const answers = req.body || {};
  const answered = questions.filter(q => answers[q.id]);
  const correct = answered.filter(q => String(answers[q.id]).trim() === q.correctAnswer).length;
  const wrong = answered.length - correct;
  const score = calculateScore(questions.length, correct, wrong);
  const percentage = calculatePercentage(score, questions.length);
  const passed = isPassed(percentage);
  await QuizResult.create({ userId: req.auth.user, score, percentage, passed });
  res.render('summary', { result: { score, percentage, passed, correct, wrong, total: questions.length } });
});

/**
 * @openapi
 * /api/quiz/submit:
 *   post:
 *     summary: Prześlij odpowiedzi na quiz
 *     description: Endpoint przyjmuje odpowiedzi studenta na quiz, oblicza wynik procentowy i status zaliczenia. Dostępne tylko dla studentów.
 *     security:
 *       - basicAuth: []
 *     requestBody:
 *       required: true
 *       content:
 *         application/json:
 *           schema:
 *             type: object
 *             properties:
 *               answers:
 *                 type: object
 *                 additionalProperties:
 *                   type: string
 *                 example:
 *                   "1": "B"
 *                   "2": "A"
 *     responses:
 *       200:
 *         description: Wynik quizu obliczony pomyślnie.
 *         content:
 *           application/json:
 *             schema:
 *               type: object
 *               properties:
 *                 score:
 *                   type: number
 *                 percentage:
 *                   type: number
 *                 passed:
 *                   type: boolean
 *                 correct:
 *                   type: integer
 *                 wrong:
 *                   type: integer
 *                 total:
 *                   type: integer
 */
router.post('/api/quiz/submit', auth, requireRole('STUDENT'), async (req, res) => {
  const questions = await Question.findAll();
  const answers = (req.body && req.body.answers) || {};
  const answered = questions.filter(q => answers[q.id]);
  const correct = answered.filter(q => String(answers[q.id]).trim() === q.correctAnswer).length;
  const wrong = answered.length - correct;
  const score = calculateScore(questions.length, correct, wrong);
  const percentage = calculatePercentage(score, questions.length);
  const passed = isPassed(percentage);
  await QuizResult.create({ userId: req.auth.user, score, percentage, passed });
  res.json({ score, percentage, passed, correct, wrong, total: questions.length });
});

module.exports = router;

File: e-learning-platform/python/routers/quiz.py
from fastapi import APIRouter, Depends, Request, Form
from fastapi.responses import HTMLResponse
from fastapi.templating import Jinja2Templates
from sqlalchemy.orm import Session
from models import SessionLocal, Question, QuizResult
from schemas import QuizSubmitRequest, QuizResultResponse
from services.score_calculator import calculate_score, calculate_percentage, is_passed
from routers.auth import get_current_user, require_student

router = APIRouter()
templates = Jinja2Templates(directory="templates")


def get_db():
    db = SessionLocal()
    try:
        yield db
    finally:
        db.close()


@router.get("/quiz", response_class=HTMLResponse)
def show_quiz(request: Request, db: Session = Depends(get_db), user: str = Depends(get_current_user)):
    questions = db.query(Question).all()
    return templates.TemplateResponse("quiz.html", {"request": request, "questions": questions})


@router.post("/quiz/submit", response_class=HTMLResponse)
async def submit_quiz(request: Request, db: Session = Depends(get_db), user: str = Depends(require_student)):
    form = await request.form()
    questions = db.query(Question).all()
    form_dict = dict(form)
    answered = list(filter(lambda q: str(q.id) in form_dict and form_dict[str(q.id)], questions))
    correct = sum(map(lambda q: int(form_dict[str(q.id)] == q.correct_answer), answered))
    wrong = len(answered) - correct
    score = calculate_score(len(questions), correct, wrong)
    percentage = calculate_percentage(score, len(questions))
    passed = is_passed(percentage)
    result = QuizResult(user_id=user, score=score, percentage=percentage, passed=passed)
    db.add(result)
    db.commit()
    resp = QuizResultResponse(score=score, percentage=percentage, passed=passed, correct=correct, wrong=wrong, total=len(questions))
    return templates.TemplateResponse("summary.html", {"request": request, "result": resp})


@router.post("/api/quiz/submit")
def api_submit_answers(req: QuizSubmitRequest, db: Session = Depends(get_db), user: str = Depends(require_student)):
    questions = db.query(Question).all()
    answers_dict = req.answers or {}
    answered = list(filter(lambda q: str(q.id) in answers_dict and answers_dict[str(q.id)], questions))
    correct = sum(map(lambda q: int(answers_dict[str(q.id)] == q.correct_answer), answered))
    wrong = len(answered) - correct
    score = calculate_score(len(questions), correct, wrong)
    percentage = calculate_percentage(score, len(questions))
    passed = is_passed(percentage)
    result = QuizResult(user_id=user, score=score, percentage=percentage, passed=passed)
    db.add(result)
    db.commit()
    return QuizResultResponse(score=score, percentage=percentage, passed=passed, correct=correct, wrong=wrong, total=len(questions))
